### DISCUSSION TITLE
Amankumar Singh - GitHub Externship Proposal

### DISCUSSION THREAD

**Discussion by Amansingh1202** [hoppscotch/hoppscotch#1630] [Q&A]
**Application number:** 4-05_Ama793_tfa_258

**Full name:** Amankumar Singh

**E-mail:** aks001235@gmail.com

**GitHub username:** Amansingh1202

**Prior experience:**
- [x] HTML
- [x] CSS
- [ ] SCSS
- [ ] Tailwind CSS
- [x] JavaScript
- [ ] TypeScript
- [x] Vue.js
- [x] Nuxt.js
- [ ] NestJS
- [ ] Firebase
- [ ] Prisma
- [ ] GraphQL
- [x] PostgreSQL
- [x] Go lang

---

**Proposal type:**
- [x] Existing feature request
- [ ] New feature proposal
- [ ] Refactor existing feature

**Proposal title:** Mock Server

**Proposal description:** 
The user can enter details of the API using the UI.Hoppscotch will process the user input and create an OpenAPI json or yaml file.This file will be used by the server to create mock data and would also provide fake endpoints to access and manipulate the data.

**Proposal roadmap:**
1)Create an UI for the user to enter details about the API.
2)Generate OpenAPI json or yaml file according to the specified user input.
3)Use **open-api-mocker** to generate mock data.Open api mocker uses faker.js to generate mock data.
4)Create endpoints to access and manipulate this generated data using **json-server**.

**Comment by AndrewBastin** [comment]
You might want to expand on this proposal a bit more ;)
Give a breakdown into how you are planning to implement it, the technologies and the architecture you are planning to use etc.

**Reply by Amansingh1202** [reply]
This is my detailed plan for the proposed feature:
1) Create a form using Vue for the users to enter the details of the API such as the field names and the type of the fields(we will have some predefined type fields like Email, GUID, Country Name, etc). Users can also enter how much data they want to generate.
2) After the user will submit the data, we will generate an OpenAPI JSON or YAML file or we can just create a simple JSON schema file(depends on which package we are using to create the fake data). We will convert the user data into the required fields of the file and will add some extra content in the file as per the requirements. The logic of the file generation can be written in javascript or any package can be used to generate JSON schema for us.
3)Then we will use an open-api-mocker that generates mock data from OpenAPI specifications. For using open-api-mocker we will need to add x-faker fields in the OpenAPI file so we will make the necessary changes in the file. Also if we choose to generate a simple JSON schema file we can use json-schema-faker to generate fake data. Use of json-schema-faker will be preferred because it is easy to implement.
4) Both the mentioned packages will generate a JSON response which can be served using json-server.
